chan.h

```diff
@@ -11,45 +11,71 @@ using std::mutex;
 using std::queue;
 using std::unique_lock;
 
-template <typename T>
-class Chan {
-    queue<T> que;
-    mutable mutex mtx;
-    condition_variable cond;
+template <typename T> class Chan {
+  // buffered queue properties
+  queue<T> que;
+  mutable mutex mtx;
+  condition_variable r_cond;
+  condition_variable w_cond;
 
-    const size_t quota;
-    size_t passed;
+  int r_waiting;
```

Comment:
We dont need this waiting counter, just notify_all, it works. check it here http://en.cppreference.com/w/cpp/thread/condition_variable/notify_all